File: firmware/test/fsm_test.cpp
// Testes da máquina de estados, rodando no PC (sem ESP32).
// Execução: pnpm test:fsm
#include <cstdio>
#include <cstdlib>
#include "occupancy_fsm.h"

static int falhas = 0;
#define CHECK(cond, msg)                                   \
  do {                                                     \
    if (!(cond)) { std::printf("FALHOU: %s\n", msg); ++falhas; } \
    else { std::printf("ok: %s\n", msg); }                 \
  } while (0)

// Simula 'ms' milissegundos em passos de 100 ms com o PIR fixo.
// Itera por número de passos (e não "t < fim"): comparar instantes absolutos
// quebra quando o contador estoura — o mesmo erro que a FSM evita.
static void rodar(FsmState& s, const FsmConfig& c, bool pir, uint32_t& t, uint32_t ms) {
  for (uint32_t passos = ms / 100; passos > 0; --passos, t += 100) fsmStep(s, c, pir, t);
}

int main() {
  // --- Modelo da Figura 3 (sem janela de confirmação) ---
  {
    const FsmConfig c{30000, 0, 1};
    FsmState s;
    uint32_t t = 0;
    rodar(s, c, false, t, 1000);
    CHECK(s.estado == Ocupacao::DESOCUPADO, "fig3: inicia desocupado");
    rodar(s, c, true, t, 500);
    CHECK(s.estado == Ocupacao::OCUPADO, "fig3: movimento -> OCUPADO");
    rodar(s, c, false, t, 29000);
    CHECK(s.estado == Ocupacao::OCUPADO, "fig3: mantém ocupado antes de t_ocupado");
    rodar(s, c, false, t, 2000);
    CHECK(s.estado == Ocupacao::DESOCUPADO, "fig3: desocupa após t_ocupado");
  }

  // --- Com janela de confirmação (2 pulsos em 10 s) ---
  {
    const FsmConfig c{30000, 10000, 2};
    FsmState s;
    uint32_t t = 0;
    rodar(s, c, true, t, 300);
    CHECK(s.estado == Ocupacao::CONFIRMANDO, "janela: 1º pulso -> CONFIRMANDO");
    rodar(s, c, false, t, 11000);
    CHECK(s.estado == Ocupacao::DESOCUPADO, "janela: pulso isolado descartado (falso positivo)");

    rodar(s, c, true, t, 300);
    rodar(s, c, false, t, 2000);
    rodar(s, c, true, t, 300);
    CHECK(s.estado == Ocupacao::OCUPADO, "janela: 2 pulsos na janela -> OCUPADO");
  }

  // --- Estouro do millis() (~49,7 dias) ---
  {
    const FsmConfig c{30000, 0, 1};
    FsmState s;
    uint32_t t = 0xFFFFFFFFu - 1000;  // 1 s antes do estouro
    rodar(s, c, true, t, 300);
    rodar(s, c, false, t, 10000);  // atravessa o estouro
    CHECK(s.estado == Ocupacao::OCUPADO, "overflow: não desocupa antes da hora");
    rodar(s, c, false, t, 25000);
    CHECK(s.estado == Ocupacao::DESOCUPADO, "overflow: desocupa no tempo certo");
  }

  std::printf("\n%s (%d falha(s))\n", falhas ? "FALHOU" : "TODOS OK", falhas);
  return falhas ? EXIT_FAILURE : EXIT_SUCCESS;
}

File: apps/backend/src/simulador/fsm.test.ts
// Porte dos 9 casos de firmware/test/fsm_test.cpp, com os mesmos nomes e valores.
// É a prova de que o modo Demo se comporta como o firmware.
import { describe, expect, it } from 'vitest'
import { fsmStep, novoFsmState, type FsmConfig, type FsmState } from './fsm'

// Relógio de teste: `t` avança em passos de 100 ms com aritmética uint32,
// como o millis() do ESP32.
type Relogio = { t: number }

// Simula 'ms' milissegundos em passos de 100 ms com o PIR fixo.
// Itera por número de passos (e não "t < fim"): comparar instantes absolutos
// quebra quando o contador estoura — o mesmo erro que a FSM evita.
function rodar(s: FsmState, c: FsmConfig, pir: boolean, relogio: Relogio, ms: number): void {
  for (let passos = Math.floor(ms / 100); passos > 0; --passos, relogio.t = (relogio.t + 100) >>> 0) {
    fsmStep(s, c, pir, relogio.t)
  }
}

describe('Modelo da Figura 3 (sem janela de confirmação)', () => {
  const c: FsmConfig = { tOcupadoMs: 30000, janelaConfMs: 0, pulsosConf: 1 }
  const s = novoFsmState()
  const relogio: Relogio = { t: 0 }

  it('fig3: inicia desocupado', () => {
    rodar(s, c, false, relogio, 1000)
    expect(s.estado).toBe('DESOCUPADO')
  })

  it('fig3: movimento -> OCUPADO', () => {
    rodar(s, c, true, relogio, 500)
    expect(s.estado).toBe('OCUPADO')
  })

  it('fig3: mantém ocupado antes de t_ocupado', () => {
    rodar(s, c, false, relogio, 29000)
    expect(s.estado).toBe('OCUPADO')
  })

  it('fig3: desocupa após t_ocupado', () => {
    rodar(s, c, false, relogio, 2000)
    expect(s.estado).toBe('DESOCUPADO')
  })
})

describe('Com janela de confirmação (2 pulsos em 10 s)', () => {
  const c: FsmConfig = { tOcupadoMs: 30000, janelaConfMs: 10000, pulsosConf: 2 }
  const s = novoFsmState()
  const relogio: Relogio = { t: 0 }

  it('janela: 1º pulso -> CONFIRMANDO', () => {
    rodar(s, c, true, relogio, 300)
    expect(s.estado).toBe('CONFIRMANDO')
  })

  it('janela: pulso isolado descartado (falso positivo)', () => {
    rodar(s, c, false, relogio, 11000)
    expect(s.estado).toBe('DESOCUPADO')
  })

  it('janela: 2 pulsos na janela -> OCUPADO', () => {
    rodar(s, c, true, relogio, 300)
    rodar(s, c, false, relogio, 2000)
    rodar(s, c, true, relogio, 300)
    expect(s.estado).toBe('OCUPADO')
  })
})

describe('Estouro do millis() (~49,7 dias)', () => {
  const c: FsmConfig = { tOcupadoMs: 30000, janelaConfMs: 0, pulsosConf: 1 }
  const s = novoFsmState()
  const relogio: Relogio = { t: 0xffffffff - 1000 } // 1 s antes do estouro

  it('overflow: não desocupa antes da hora', () => {
    rodar(s, c, true, relogio, 300)
    rodar(s, c, false, relogio, 10000) // atravessa o estouro
    expect(s.estado).toBe('OCUPADO')
  })

  it('overflow: desocupa no tempo certo', () => {
    rodar(s, c, false, relogio, 25000)
    expect(s.estado).toBe('DESOCUPADO')
  })
})
